File: javascript/protein-translation/protein-translation.js
const convertion = {
  'AUG': 'Methionine',
  'UUU': 'Phenylalanine',
  'UUC': 'Phenylalanine',
  'UUA': 'Leucine',
  'UUG': 'Leucine',
  'UCU': 'Serine',
  'UCC': 'Serine',
  'UCA': 'Serine',
  'UCG': 'Serine',
  'UAU': 'Tyrosine',
  'UAC': 'Tyrosine',
  'UGU': 'Cysteine',
  'UGC': 'Cysteine',
  'UGG': 'Tryptophan'//,
  // 'UAA': 'STOP',
  // 'UAG': 'STOP',
  // 'UGA': 'STOP'
};

const stops = ['UAA', 'UAG', 'UGA'];

export default function translate(str) {
  const translation = [];
  if (str === undefined) { return translation; }

  const codons = str.split(/(...)/).filter(obj => obj);
  let stopIndex = codons.findIndex(findStop) === -1 ? codons.length : codons.findIndex(findStop);
  const toTranslate = codons.slice(0, stopIndex);

  toTranslate.forEach(codon => {
    if (convertion[codon] == undefined) { throw new Error('Invalid codon'); }
    translation.push(convertion[codon]);
  })

  return translation;
}

function findStop(codon) {
  return stops.includes(codon);
}

File: ruby/protein-translation/protein_translation.rb
class InvalidCodonError < StandardError; end

class Translation
  CONVERTION = {
    'AUG' => 'Methionine',
    'UUU' => 'Phenylalanine',
    'UUC' => 'Phenylalanine',
    'UUA' => 'Leucine',
    'UUG' => 'Leucine',
    'UCU' => 'Serine',
    'UCC' => 'Serine',
    'UCA' => 'Serine',
    'UCG' => 'Serine',
    'UAU' => 'Tyrosine',
    'UAC' => 'Tyrosine',
    'UGU' => 'Cysteine',
    'UGC' => 'Cysteine',
    'UGG' => 'Tryptophan',
    'UAA' => 'STOP',
    'UAG' => 'STOP',
    'UGA' => 'STOP'
  }

  def self.of_codon(codon)
    raise InvalidCodonError if CONVERTION[codon].nil?

    CONVERTION[codon]
  end

  def self.of_rna(strand)
    strand.scan(/.../).each_with_object([]) do |codon, output|
      return output if CONVERTION[codon] == 'STOP'

      output << of_codon(codon)
    end
  end
end
